### ISSUE TITLE
Run time analysis of charikarDicts

### ISSUE THREAD

**Issue by joseduc10** [polinapolina/Charikar-densest-subgraph#1]
Why is the running time of charikarDicts O(mn)?

The outer loop iterates through each node once. The inner for loop iterates through
the neighbors of the "chosen" node in a given iteration. Therefore, over *all* iterations
the inner loop is repeated O(m) times. It seems to me the running time is O(m + n).

Am I missing something?

**Comment by polinapolina** [comment]
Right, O(mn) is too pessimistic.

If calculate precisely, then, yes, total number of iterations over neighbors gives O(m).
But total iterations over nodes to find a node with the smallest degree gives O(n(n+1)/2)=O(n^2) as a sum of arithmetic progression n + (n-1) + .. +1. Please correct me, if I'm wrong here.
As a result, O(m+n^2).

To implement the algorithm in O(m+n) one needs to keep another dictionary on top, such that key is an integer from 0 to n and value is a list of nodes, which have degree equal to keys value. This implementation is described in original paper by Moses Charikar (Greedy Approximation Algorithms for finding Dense Components in a Graph).

**Comment by joseduc10** [comment]
Yes, you are right. I forgot to count the loop for the smallest degree computation. The running time should be O(m + n^2), or just O(n^2).

In case you are interested, I forked your repo and added the implementation described by Charikar: https://github.com/joseduc10/Charikar-densest-subgraph . I also noticed that deep-copying the Graph object is costly. It is better to just maintain the graph structure in auxiliary dictionaries.

Thank you very much for the clarification.

**Comment by joseduc10** [comment]
By the way, I read your KDD paper "Event Detection in Activity Networks" and found it very interesting. I am a graduate student at Virginia Tech, and I am working on problems on anomaly detection on  networks and Steiner connectivity. I an wondering if you plan on extending your KDD work.

**Comment by polinapolina** [comment]
Cool, very neat implementation!

Thank you for a nice comment on our KDD paper. We plan to extend it by adding explicit temporal component into the model. However, right now we are interested in exploring temporal networks as interaction networks, like it was done in our paper "Discovering dynamic communities
in interaction networks".

Anomaly detection on networks is one of my great interests, as it is non-trivial to combine temporal and graph-structural components. Do you have any work already done in this field?
